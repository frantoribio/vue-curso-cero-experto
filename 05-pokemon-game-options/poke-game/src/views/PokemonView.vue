<template>
  <h1 v-if="!pokemon">Espere por favor...</h1>
  <div v-else>
    <h1>¿Quién es este Pokemon</h1>
    <PokemonPicture
      :pokemon-id="pokemon.id"
      :show-pokemon="showPokemon"
    />
    <PokemonOptions
      :pokemons="pokemons"
      @selection-pokemon="checkAnswer"
    />
    <!-- Se puede usar templete para no mostrar div -->
    <div
      v-if="showAnswer"
      class="fade-in"
    >
      <h2>{{ message }}</h2>
      <button @click="newGame">Jugar otra vez</button>
    </div>
    <h4>Estadisticas:</h4>
    <h5>Partidas: {{ estadisticas.partidas }}</h5>
    <h5>Aciertos: {{ estadisticas.victorias }}</h5>
    <h5>Derrotas: {{ estadisticas.derrotas }}</h5>
  </div>
</template>

<script>
  import PokemonPicture from '@/components/PokemonPicture.vue'
  import PokemonOptions from '@/components/PokemonOptions.vue'
  import getPokemons from '@/services/pokeservice'
  import { mapState } from 'pinia'
  import { PokeStore } from '@/stores/pokestore'

  export default {
    name: 'PokemonView',
    components: {
      PokemonPicture,
      PokemonOptions,
    },
    data() {
      return {
        pokemons: [],
        pokemon: null,
        showPokemon: false,
        showAnswer: false,
        message: '',
      }
    },
    methods: {
      async setPokemons() {
        this.pokemons = await getPokemons()
        this.pokemon = this.pokemons[Math.floor(Math.random() * this.pokemons.length)]
      },
      checkAnswer(pokemonId) {
        if (pokemonId === this.pokemon.id) {
          this.showPokemon = true
          this.showAnswer = true
          this.message = `¡Correcto! es ${this.pokemon.name}`
          this.estadisticas.victorias++
        } else {
          this.showPokemon = false
          this.showAnswer = true
          this.message = `¡Incorrecto! es ${this.pokemon.name}`
          this.estadisticas.derrotas++
        }
      },
      newGame() {
        this.pokemon = null
        this.pokemons = []
        this.showAnswer = false
        this.showPokemon = false
        this.message = ''
        this.setPokemons()
        this.estadisticas.partidas++
      },
    },
    mounted() {
      this.setPokemons()
    },
    computed: {
      ...mapState(PokeStore, ['estadisticas']),
    },
  }
</script>

<style lang="scss" scoped></style>
