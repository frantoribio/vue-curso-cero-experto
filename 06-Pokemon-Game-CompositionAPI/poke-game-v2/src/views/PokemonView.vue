<template>
  <h1 v-if="!pokemon">Espere por favor...</h1>
  <div v-else>
    <h1>¿Quién es este Pokemon</h1>
    <PokemonPicture
      data-testid="PokemonPicture"
      :pokemon-id="pokemon.id"
      :show-pokemon="showPokemon"
    />
    <PokemonOptions
      data-testid="PokemonOptions"
      :pokemons="pokemons"
      @selection-pokemon="checkAnswer"
    />
    <!-- Se puede usar templete para no mostrar div -->
    <div
      v-if="showAnswer"
      class="fade-in"
    >
      <h2>{{ message }}</h2>
      <button @click="newGame">Jugar otra vez</button>
    </div>
    <h4>Estadisticas:</h4>
    <h5>Partidas: {{ estadisticas.partidas }}</h5>
    <h5>Aciertos: {{ estadisticas.victorias }}</h5>
    <h5>Derrotas: {{ estadisticas.derrotas }}</h5>
  </div>
</template>

<script>
  import PokemonPicture from '@/components/PokemonPicture.vue'
  import PokemonOptions from '@/components/PokemonOptions.vue'
  import getPokemons from '@/services/pokeservice'
  import { PokeStore } from '@/stores/pokestore'
  import { ref, onMounted, watch, computed } from 'vue'

  export default {
    name: 'PokemonView',
    components: {
      PokemonPicture,
      PokemonOptions,
    },
    //  setup(props, context)
    setup() {
      // Mis objetos reactivos!!!
      const pokemons = ref([])
      const pokemon = ref(null)
      const showPokemon = ref(false)
      const showAnswer = ref(false)
      const message = ref('')

      // Para o poner ejemplo con el watcher
      let idPokemon = 0 // no va a ser reactivo

      // La stores
      const pokeStore = PokeStore()

      // Ciclo de vida on Mounted
      onMounted(() => {
        setPokemons()
      })

      const setPokemons = async () => {
        pokemons.value = await getPokemons()
        pokemon.value = pokemons.value[Math.floor(Math.random() * pokemons.value.length)]
      }

      const newGame = () => {
        //pokemon.value = null
        pokemons.value = []
        showAnswer.value = false
        showPokemon.value = false
        message.value = ''
        setPokemons()
        pokeStore.estadisticas.partidas++
      }

      const checkAnswer = (pokemonId) => {
        if (pokemonId === pokemon.value.id) {
          showPokemon.value = true
          showAnswer.value = true
          message.value = `¡Correcto! es ${pokemon.value.name}`
          pokeStore.estadisticas.victorias++
        } else {
          showPokemon.value = false
          showAnswer.value = true
          message.value = `¡Incorrecto! es ${pokemon.value.name}`
          pokeStore.estadisticas.derrotas++
        }
      }

      // Computed, es un ejemplo pero lo cambia cada vez que lo lea!!!
      const pokemonData = computed(() => {
        return `Pokemon: ${pokemon.value.name} tiene el id:(${pokemon.value.id})`
      })

      // watcher
      watch(pokemon, (newPokemon) => {
        // Solo por hacer un ejemplo
        idPokemon = newPokemon.id
        //console.log(idPokemon)
      })

      // Devulvemos un objeto con los datos de la vista
      return {
        pokemons,
        pokemon,
        showPokemon,
        showAnswer,
        message,
        newGame,
        checkAnswer,
        estadisticas: pokeStore.estadisticas,
        // Para testear, la hago publica
        setPokemons,
        idPokemon,
        pokemonData,
      }

      // Para testear si queremos, exponemos
    },
  }
</script>

<style lang="scss" scoped></style>
