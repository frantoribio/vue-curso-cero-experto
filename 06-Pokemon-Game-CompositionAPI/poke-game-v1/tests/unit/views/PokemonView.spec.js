import { describe, test, expect, vi } from 'vitest'
import { shallowMount } from '@vue/test-utils'
import PokemonView from '@/views/PokemonView.vue'
import { pokemons } from '../mocks/pokemons.mocks'
import { createTestingPinia } from '@pinia/testing'

describe('Pokemon View Component', () => {
  test('Debe hacer match con el snapshot cuando cargan los pokemons', () => {
    // Vamos a montar el stubs!!!
    const wrapper = shallowMount(PokemonView, {
      global: {
        plugins: [
          createTestingPinia({
            createSpy: vi.fn,
          }),
        ],
      },
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  test('Debe llamar a setPokemons() al montarse', () => {
    const wrapper = shallowMount(PokemonView, {
      mounted: vi.fn(), // De esta manera mockeamos el onMounted
      global: {
        plugins: [
          createTestingPinia({
            createSpy: vi.fn,
          }),
        ],
      },
    })
    //console.log(wrapper)
    // Creamos el spia
    const setPokemonsSpy = vi.spyOn(wrapper.vm, 'setPokemons')

    wrapper.vm.setPokemons()

    expect(setPokemonsSpy).toHaveBeenCalled()
  })

  test('debe mostrar los compo PokemonPicture y PokemonOptions', async () => {
    const wrapper = shallowMount(PokemonView, {
      mounted: vi.fn(), // De esta manera mockeamos el onMounted
      global: {
        plugins: [
          createTestingPinia({
            createSpy: vi.fn,
          }),
        ],
      },
    })

    // Asignamos las variables para que sean las que queremos
    // Ya no podemos usar data porque no lo tenemos!!!
    // console.log(wrapper.vm.$data)
    // No es lo mismo que
    // console.log(wrapper.vm)
    // Estamos usando composition
    wrapper.vm.pokemons = pokemons
    wrapper.vm.pokemon = pokemons[0]
    await wrapper.vm.$nextTick() // Esperamos cambios de renderizado
    // console.log(wrapper.vm.$data)
    // Mejor
    const pokemonPicture = wrapper.find('[data-testid="PokemonPicture"]')
    const pokemonOptions = wrapper.find('[data-testid="PokemonOptions"]')

    expect(pokemonPicture.exists()).toBe(true)
    expect(pokemonOptions.exists()).toBe(true)

    expect(pokemonPicture.isVisible()).toBe(true)
    expect(pokemonOptions.isVisible()).toBe(true)

    expect(pokemonPicture.attributes('pokemonid')).toBeTruthy() // Al ser computer la leo aleatoria, por eso debe estar
    expect(pokemonOptions.attributes('pokemons')).toBeTruthy()

    //console.log(pokemonPicture.html())
  })

  test('pruebas con check answer', async () => {
    // Montamos un componente igual que el que queremos probar
    // Aquí comenzamos!!!!
    const wrapper = shallowMount(PokemonView, {
      mounted: vi.fn(), // De esta manera mockeamos el onMounted
      global: {
        plugins: [
          createTestingPinia({
            createSpy: vi.fn,
          }),
        ],
      },
    })

    wrapper.vm.pokemons = pokemons
    wrapper.vm.pokemon = pokemons[0]
    await wrapper.vm.$nextTick()
    //console.log(wrapper.vm.pokemon)

    // Probamos el incorrecto
    await wrapper.vm.checkAnswer(5)
    expect(wrapper.find('h2').exists()).toBe(true)
    expect(wrapper.find('h2').text()).toBe(`¡Incorrecto! es ${pokemons[0].name}`)
    expect(wrapper.vm.showPokemon).toBe(false)
    expect(wrapper.vm.showAnswer).toBe(true)
    expect(wrapper.vm.estadisticas.partidas).toBe(1)
    expect(wrapper.vm.estadisticas.derrotas).toBe(1)

    // Probamos el correcto
    await wrapper.vm.checkAnswer(1)
    expect(wrapper.find('h2').exists()).toBe(true)
    expect(wrapper.find('h2').text()).toBe(`¡Correcto! es ${pokemons[0].name}`)
    expect(wrapper.vm.showPokemon).toBe(true)
    expect(wrapper.vm.showAnswer).toBe(true)
    expect(wrapper.vm.estadisticas.victorias).toBe(1)
  })

  test('Botón se ha pulsado', async () => {
    const wrapper = shallowMount(PokemonView, {
      mounted: vi.fn(), // De esta manera mockeamos el onMounted
      global: {
        plugins: [
          createTestingPinia({
            createSpy: vi.fn,
          }),
        ],
      },
    })

    // cargamos los datos
    wrapper.vm.pokemons = pokemons
    wrapper.vm.pokemon = pokemons[0]
    await wrapper.vm.$nextTick()

    // probamos a pulsar una
    await wrapper.vm.checkAnswer(1)

    const myButton = wrapper.find('button')
    // disparameos el evento
    await myButton.trigger('click')
    // Estadisticas aumenta la partida
    expect(wrapper.vm.estadisticas.partidas).toBe(2)
  })

  test('Estadisticas', async () => {
    const wrapper = shallowMount(PokemonView, {
      mounted: vi.fn(), // De esta manera mockeamos el onMounted
      global: {
        plugins: [
          createTestingPinia({
            createSpy: vi.fn,
          }),
        ],
      },
    })

    // cargamos los datos
    wrapper.vm.pokemons = pokemons
    wrapper.vm.pokemon = pokemons[0]
    await wrapper.vm.$nextTick()

    // Probamos el incorrecto y
    await wrapper.vm.checkAnswer(5)
    await wrapper.vm.checkAnswer(1)
    expect(wrapper.vm.estadisticas.partidas).toBe(1)
    expect(wrapper.vm.estadisticas.derrotas).toBe(1)
    expect(wrapper.vm.estadisticas.victorias).toBe(1)

    const [p1, a1, d1] = wrapper.findAll('h5')
    expect(p1.text()).toBe('Partidas: 1')
    expect(a1.text()).toBe('Aciertos: 1')
    expect(d1.text()).toBe('Derrotas: 1')

    // PulSAMOS EL BOTÓN
    const myButton = wrapper.find('button')
    await myButton.trigger('click')

    expect(wrapper.vm.estadisticas.partidas).toBe(2)
    expect(wrapper.vm.estadisticas.derrotas).toBe(1)
    expect(wrapper.vm.estadisticas.victorias).toBe(1)
  })

  test('Computed y Watcher', async () => {
    const wrapper = shallowMount(PokemonView, {
      mounted: vi.fn(), // De esta manera mockeamos el onMounted
      global: {
        plugins: [
          createTestingPinia({
            createSpy: vi.fn,
          }),
        ],
      },
    })

    // cargamos los datos
    wrapper.vm.pokemons = pokemons
    wrapper.vm.pokemon = pokemons[0]
    await wrapper.vm.$nextTick()

    expect(wrapper.vm.pokemonData).toBe(
      `Pokemon: ${pokemons[0].name} tiene el id:(${pokemons[0].id})`
    )

    expect(wrapper.vm.idPokemon).toBe(0)
  })
})
