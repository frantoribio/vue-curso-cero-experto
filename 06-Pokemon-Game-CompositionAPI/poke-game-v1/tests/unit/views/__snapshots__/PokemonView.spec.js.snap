// Vitest Snapshot v1

exports[`Pokemon View Component > Debe hacer match con el snapshot cuando cargan los pokemons 1`] = `
<h1>
  Espere por favor...
</h1>
`;
